### ISSUE TITLE
Think about caching queries

### ISSUE THREAD

**Issue by alexkreidler** [oxigraph/oxigraph#95]
The query from issue #94 takes a while (18 seconds) to run on 36M triples which take up 5.7 GB on disk. When I run it again, it takes a similar time (17 seconds).

Adding an option to cache query responses might help. Some other options with this:

- Using the parsed version of the query to create the cache key
- Possibly caching intermediate steps (e.g. a small set of frequently accessed triples)

Of course, caching adds a whole level of complexity and creates issues with invalidation and the consistency of the system. But it is probably good to think about for the future!

**Comment by Tpt** [comment]
Caching is an hard problem indeed. I prefer to focus first on query optimizations that would also help for new queries and does not require to deal with invalidation. The query evaluator is not much optimized yet, there are still a lot of low hanging fruits (like #94) to do first imho.

**Comment by Tpt** [comment]
A relevant paper: [Semantics and Canonicalisationof SPARQL 1.1](http://semantic-web-journal.net/system/files/swj2757.pdf)

**Comment by pangoSE** [comment]
What about no cache-as-a-feature? Keeping it small, simple, fast and efficient. The web frontend can have something like nginx in front of it to take care of caching. See https://en.wikipedia.org/wiki/Web_cache#Server-side_software

**Comment by dpriskorn** [comment]
What about no cache-as-a-feature? Keeping it small, simple, fast and efficient. The web frontend can have something like nginx in front of it to take care of caching. See https://en.wikipedia.org/wiki/Web_cache#Server-side_software

**Comment by alexkreidler** [comment]
I understand the desire to reduce the surface area that the database is responsible for. However the paper linked above describes a canonicalization algorithm that turns two SPARQL queries that are logically equivalent into identical strings. 

Another issue for example is a situation where one query contains a subgraph that has already been computed. If Oxigraph did caching at the application level, it could run the normalization from the paper, recognize that a subgraph is in a cache map, retrieve the cached data for the subgraph, and then only have to process the rest of the query.

The issue with "dumb" caching is such logically equivalent or subgraph queries wouldn't be cached. Of course, this might not be an issue depending on the workload, and this definitely isn't a high priority feature at this stage, but I think it's worthwhile to consider for the future.

**Comment by Tpt** [comment]
Yes, I agree that making Oxigraph as fast as possible without caching is the main goal. Cache would be only an optional workaround for intrinsically slow queries. If I have to prioritize, I would say this is a "post 1.0" feature.

However, if we want to have some query cache, Oxigraph is probably the best place for that because we could easily implement in it some kind of query canonicalization and reuse already cached subqueries/subgraphs. This might improve a lot the cache-hit ratio compared to a simple web frontend.

**Comment by ktk** [comment]
The nice thing about SPARQL is that we can use any HTTP caching service for it in front. As long as you send the same SPARQL query (as in the same hash) it would immediately return the result from the previous query. We use this in production and once you figure out how Varnish needs to be configured & send the right headers in the HTTP response, this works flawlessly. Tricky part is that Varnish is only caching GET requests by default so you need a VCL synth config for POST requests but it's not that hard to do. Will document that in our blog in the next weeks so others can see what we did.
